src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java

```diff
@@ -2905,6 +2908,187 @@ public void visitTypeCast(JCTypeCast tree) {
+                VarSymbol dollar_s = new VarSymbol(FINAL | SYNTHETIC,
+                        names.fromString("tmp" + tree.pos + this.target.syntheticNameChar()),
+                        tree.expr.type,
+                        currentMethodSym);
+
+                JCStatement var = make.at(tree.pos()).VarDef(dollar_s, instanceOfExpr).setType(dollar_s.type);
+
+                if (tree.expr.type.isReference()) {
+                    JCExpression nullCheck = makeBinary(NE,
+                            make.Ident(dollar_s),
+                            makeNull());
+
+                    if (types.checkUnconditionallyExact(types.unboxedType(tree.expr.type), tree.pattern.type)) {
+                        exactnessCheck = make.Literal(BOOLEAN, 1).setType(syms.booleanType);                                          // emit no exactness check
+                    } else {
+                        // if expression type is Byte, Short, Integer, ...
+                        // an unboxing conversion followed by a widening primitive conversion
+                        if (types.unboxedType(tree.expr.type).isPrimitive()) {
+                            exactnessCheck = getExactnessCheck(tree, boxIfNeeded(make.Ident(dollar_s), types.unboxedType(tree.expr.type))); // emit the exactness call
+                        } else {
+                            // if expression type is a supertype: Number, ..
+                            // a narrowing reference conversion followed by an unboxing conversion
+                            exactnessCheck = make.at(tree.pos()).TypeTest(tree.expr, make.Type(types.boxedClass(tree.pattern.type).type)).setType(syms.booleanType);;
+                        }
+                    }
+
+                    JCBinary nullCheckFollowedByExactnessCheckCall = makeBinary(AND,
+                            nullCheck,
+                            exactnessCheck);
+
+                    exactnessCheck = make.LetExpr(List.of(var), nullCheckFollowedByExactnessCheckCall)
+                            .setType(syms.booleanType);
+                } else {
+                    // rewrite instanceof if expr : primitive
+                    // int v = ...
+                    // if (v instanceof float)
+                    // =>
+                    // if (let tmp$123 = v; ExactnessChecks.int_float(tmp$123))
+                    JCIdent argument = make.Ident(dollar_s);
+
+                    JCExpression exactnessCheckCall =
+                            getExactnessCheck(tree, argument);
+
+                    exactnessCheck = make.LetExpr(List.of(var), exactnessCheckCall)
+                            .setType(syms.booleanType);
+                }
+            }
+
+            result = exactnessCheck;
+        }
+        else {
+            tree.expr = translate(tree.expr);
+            tree.pattern = translate(tree.pattern);
+            result = tree;
+        }
+    }
+
+    static class TypePairs {
```

Comment:
I believe this could be a record between two TypeSymbols? Then you could have an additional constructor that takes two types and retrieves their `tsym` and call the canonical constructor (and then you can drop equals/hashcode).

Reply:
Done and simplified https://github.com/openjdk/jdk/pull/15638/commits/ef98c040c909981077936c6ff4280b6da04b8c24. Do you think we can simplify further (or benefit at all) by also eliminating duplication between SwitchBootstrap and Lower?

Reply:
I don't think there's an easy easy way to deduplicate the code, as the code occurs in two different modules, so it's messy (as bootstrap javac compilation cannot depend on stuff that is in the latest latest JDK, as it has to compile against JDK N-1, and, on the other hand, java.base cannot depend on any other modules). So, leave it as is, maybe add a comment that the code should stay in sync :-)